src/core/shared_memory_manager.cc

```diff
@@ -0,0 +1,244 @@
+
+Status
+open_shm_region(const std::string& shm_key, int* shm_fd)
+{
+  // get shared memory region descriptor
+  *shm_fd = shm_open(shm_key.c_str(), O_RDONLY, S_IRUSR | S_IWUSR);
+  if (*shm_fd == -1) {
+    return Status(
+        RequestStatusCode::INTERNAL, "Unable to get shared memory descriptor");
+  }
+
+  return Status::Success;
+}
+
+Status
+get_shm_addr(
+    const int shm_fd, const size_t offset, const size_t byte_size,
+    void** shm_addr)
+{
+  // map shared memory to process address space
+  *shm_addr = mmap(NULL, byte_size, PROT_WRITE, MAP_SHARED, shm_fd, offset);
+  if (*shm_addr == MAP_FAILED) {
+    return Status(
+        RequestStatusCode::INTERNAL, "Unable to process address space");
+  }
+
+  return Status::Success;
+}
+
+Status
+shm_close(int shm_fd)
+{
+  int tmp = close(shm_fd);
+  if (tmp == -1) {
+    return Status(
+        RequestStatusCode::INTERNAL, "Unable to close shared memory region");
+  }
+
+  return Status::Success;
+}
+
+Status
+mmap_cleanup(void* shm_addr, size_t byte_size)
+{
+  int tmp_fd = munmap(shm_addr, byte_size);
+  if (tmp_fd == -1) {
+    return Status(
+        RequestStatusCode::INTERNAL, "Unable to munmap shared memory region");
+  }
+
+  return Status::Success;
+}
+
+Status
+SharedMemoryManager::RegisterSharedMemory(
+    const std::string& name, const std::string& shm_key, const size_t offset,
+    const size_t byte_size)
+{
+  LOG_VERBOSE(1) << "Register() shared memory region: '" << name << "'";
+  Status status = Status::Success;
+
+  // check if key is in shared_memory_map_ then remove
+  if (shared_memory_map_.find(name) != shared_memory_map_.end()) {
+    UnregisterSharedMemory(name);
+  }
+
+  // register (or re-register)
+  try {
+    void* tmp_addr;
+    int shm_fd;
+
+    // don't re-open if shared memory is alreay open
+    for (auto itr = shared_memory_map_.begin(); itr != shared_memory_map_.end();
+         ++itr) {
+      if (itr->second->shm_key_ == shm_key)
+        shm_fd = itr->second->shm_fd_;
+      else {
+        RETURN_IF_ERROR(open_shm_region(shm_key, &shm_fd));
+      }
+    }
+
+    RETURN_IF_ERROR(get_shm_addr(shm_fd, offset, byte_size, &tmp_addr));
+    std::unique_ptr<SharedMemoryInfo> shm_info(new SharedMemoryInfo(
+        name, shm_key, offset, byte_size, shm_fd, tmp_addr));
+    shared_memory_map_.insert(std::make_pair(
+        name, std::unique_ptr<SharedMemoryInfo>(new SharedMemoryInfo(
+                  name, shm_key, offset, byte_size, shm_fd, tmp_addr))));
+  }
+  catch (std::exception& ex) {
+    return Status(
+        RequestStatusCode::INVALID_ARG,
+        "Unable to register shared memory region successfully.");
+  }
+
+  return Status::Success;
+}
+
+Status
+SharedMemoryManager::UnregisterSharedMemory(const std::string& name)
+{
+  LOG_VERBOSE(1) << "Unregister() '" << name << "'";
+
+  auto it = shared_memory_map_.find(name);
+  if (it != shared_memory_map_.end()) {
+    void* tmp_addr;
+    RETURN_IF_ERROR(get_shm_addr(
```

Comment:
Don't we have all the information we need from the map?

Reply:
Yes we do. fixed this